File: lox/Scanner.java
package lox;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static lox.TokenType.*; // TODO: probably shouldn't .* but Java seems not willing to let me just import the enum

class Scanner {
    private final String source;
    private final List<Token> tokens = new ArrayList<>();

    private int start = 0;   // First char of lexeme
    private int current = 0; // Offset from start to current char
    private int line = 1;

    private static final Map<String, TokenType> keywords;

    static {
        keywords = new HashMap<>();
        keywords.put("and",    AND);
        keywords.put("class",  CLASS);
        keywords.put("else",   ELSE);
        keywords.put("false",  FALSE);
        keywords.put("for",    FOR);
        keywords.put("fn",     FUNCTION);
        keywords.put("if",     IF);
        keywords.put("nil",    NIL);
        keywords.put("or",     OR);
        keywords.put("print",  PRINT);
        keywords.put("return", RETURN);
        keywords.put("super",  SUPER);
        keywords.put("this",   THIS);
        keywords.put("true",   TRUE);
        keywords.put("var",    VAR);
        keywords.put("while",  WHILE);
    }

    Scanner(String source) {
        this.source = source;
    }

    List<Token> scanTokens() {
        // Continously add tokens until we reach the end of file.
        while(!consumedAllChars()) {
            start = current;
            scanToken();
        }
        tokens.add(new Token(EOF, "", null, line));

        return tokens;
    }

    private boolean consumedAllChars() {
        return current >= source.length();
    }

    private char advance() {
        return source.charAt(current++);
    }

    private void addToken(TokenType type) {
        addToken(type, null);
    }

    private void addToken(TokenType type, Object literal) {
        String text = source.substring(start, current);
        tokens.add(new Token(type, text, literal, line));
    }

    private void scanToken() {
        char c = advance();
        switch (c) {
            case '(': addToken(LEFT_PAREN);  break;
            case ')': addToken(RIGHT_PAREN); break;
            case '{': addToken(LEFT_BRACE);  break;
            case '}': addToken(RIGHT_BRACE); break;
            case ',': addToken(COMMA);       break;
            case '.': addToken(DOT);         break;
            case '-': addToken(MINUS);       break;
            case '+': addToken(PLUS);        break;
            case ';': addToken(SEMICOLON);   break;
            case '*': addToken(STAR);        break;
            case '!':
                addToken(match('=') ? BANG_EQUAL : BANG);
                break;
            case '=':
                addToken(match('=') ? EQUAL_EQUAL : EQUAL);
                break;
            case '<':
                addToken(match('=') ? LESS_EQUAL : LESS);
                break;
            case '>':
                addToken(match('=') ? GREATER_EQUAL : GREATER);
                break;
            case '/': skipLineComment(); break;

            case ' ':
            case '\r':
            case '\t':
            // Ignore whitespace
                break;

            case '\n':
                line++;
                break;

            case '"': scanString(); break;

            default:
                if (isDigit(c)) {
                    scanNumber();
                } else if (isAlpha(c)) {
                    scanIdentifier();
                } else {
                    // TODO: continously look at further illegal tokens and bunch them into one lexical error
                    Lox.error(line, "Unexpected character: " + c);
                }
        }
    }

    private void skipLineComment() {
        // Assumes current char is '/'
        if (match('/')) {
            while (peek() != '\n' && !consumedAllChars()) {
                advance();
            }
        } else {
            addToken(SLASH);
        }
    }

    private void scanString() {
        while (peek() != '"' && !consumedAllChars()) {
            if (peek() == '\n') line++;
            advance();
        }

        if (consumedAllChars()) {
            // Must have reached EOF before closing quote
            Lox.error(line, "Unterminated string.");
        }

        advance(); // Accepting closing '"'.

        // Trim surrounding quotes,
        // NOTE: Lox doesn't support doesn't support escape sequences, possible TODO
        String value = source.substring(start + 1, current - 1);
        addToken(STRING, value);
    }

    private void scanNumber() {
        while (isDigit(peek())) {
            advance();
        }

        // Look for fractional part.
        if (peek() == '.' && isDigit(peekNext())) {
            advance(); // Conusmes '.'

            while (isDigit(peek())) {
                advance();
            }
        }

        addToken(NUMBER, Double.parseDouble(source.substring(start, current)));
    }

    private void scanIdentifier() {
        while (isAlphaNumeric(peek())) {
            advance();
        }

        String text = source.substring(start, current);
        TokenType type = keywords.get(text);
        if (type == null) {
            type = IDENTIFIER;
        }
        addToken(type);
    }

    private boolean match(char expected) {
        if (consumedAllChars()) return false;
        if (source.charAt(current) != expected) return false;

        current++;
        return true;
    }

    private char peek() {
        if (consumedAllChars()) return '\0';
        return source.charAt(current);
    }

    private char peekNext() {
        if (current + 1 >= source.length()) return '\0';
        return source.charAt(current + 1);
    }

    private boolean isDigit(char c) {
        return c >= '0' && c <= '9';
    }

    private boolean isAlpha(char c) {
        return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
    }

    private boolean isAlphaNumeric(char c) {
        return isAlpha(c) || isDigit(c);
    }

}

File: zlox/src/scanner.zig
const std = @import("std");

pub const TokenType = enum {
    // Single-character tokens
    LeftParen, RightParen, LeftBrace, RightBrace, Comma, Dot, Minus, Plus, Semicolon, Slash, Star,
    // One or two character tokens
    Bang, BangEqual, Equal, EqualEqual, Greater, GreaterEqual, Less, LessEqual,
    // Literals
    Identifier, String, Number,
    // Keywords
    And, Class, Else, False, For, Fn, If, Nil, Or, Print, Return, Super, This, True, Var, While,
    // Other
    Error, EOF
};

pub const Token = struct {
    tokenType: TokenType,
    lexeme: []const u8,
    line: u32,
};

pub const Scanner = struct {
    start: []const u8,
    current: usize,
    line: u32,

    pub fn create(source: []const u8) Scanner {
        return Scanner{
            .start = source,
            .current = 0,
            .line = 1,
        };
    }

    pub fn scanToken(self: *Scanner) Token {
        self.skipWhitespace();

        self.start = self.start[self.current..];
        self.current = 0;

        if (self.isAtEnd()) return self.makeToken(.EOF);

        const char = self.peek();
        self.advance();

        return switch (char) {
            '(' => self.makeToken(.LeftParen),
            ')' => self.makeToken(.RightParen),
            '{' => self.makeToken(.LeftBrace),
            '}' => self.makeToken(.RightBrace),
            ';' => self.makeToken(.Semicolon),
            ',' => self.makeToken(.Comma),
            '.' => self.makeToken(.Dot),
            '-' => self.makeToken(.Minus),
            '+' => self.makeToken(.Plus),
            '/' => self.makeToken(.Slash),
            '*' => self.makeToken(.Star),
            '!' => self.makeToken(if (self.match('=')) .BangEqual else .Bang),
            '=' => self.makeToken(if (self.match('=')) .EqualEqual else .Equal),
            '<' => self.makeToken(if (self.match('=')) .LessEqual else .Less),
            '>' => self.makeToken(if (self.match('=')) .GreaterEqual else .Greater),
            '"' => self.scanString(),
            else => {
                if (isAlpha(char)) return self.scanIdentifier();
                if (isDigit(char)) return self.scanNumber();
                return self.errorToken("Unexpected character.");
            }
        };
    }

    fn scanIdentifier(self: *Scanner) Token {
        while (isAlpha(self.peek()) or isDigit(self.peek())) self.advance();

        return self.makeToken(self.identifierType());
    }

    fn scanNumber(self: *Scanner) Token {
        while (isDigit(self.peek())) self.advance();

        // Look for a fractional part.
        if (self.peek() == '.' and isDigit(self.peekNext())) {
            self.advance(); // Consume `.`.

            while (isDigit(self.peek())) self.advance();
        }

        return self.makeToken(.Number);
    }

    fn scanString(self: *Scanner) Token {
        while (self.peek() != '"' and !self.isAtEnd()) {
            if (self.peek() == '\n') self.line += 1;
            self.advance();
        }

        if (self.isAtEnd()) return self.errorToken("Unterminated string.");

        // Need to consume the closing quote `"`.
        self.advance();

        return self.makeToken(.String);
    }

    fn makeToken(self: *Scanner, tokenType: TokenType) Token {
        return Token{
            .tokenType = tokenType,
            .lexeme = self.start[0..self.current],
            .line = self.line,
        };
    }

    fn errorToken(self: *Scanner, message: []const u8) Token {
        return Token{
            .tokenType = .Error,
            .lexeme = message,
            .line = self.line,
        };
    }

    fn isAtEnd(self: *Scanner) bool {
        return self.current >= self.start.len;
    }

    fn advance(self: *Scanner) void {
        self.current += 1;
    }

    fn peek(self: *Scanner) u8 {
        if (self.isAtEnd()) return 0;
        return self.start[self.current];
    }

    fn peekNext(self: *Scanner) u8 {
        if (self.current + 1 >= self.start.len) return 0;
        return self.start[self.current + 1];
    }

    fn match(self: *Scanner, expected: u8) bool {
        if (self.isAtEnd()) return false;
        if (self.peek() != expected) return false;

        self.current += 1;

        return true;
    }

    fn checkKeyword(self: *Scanner, startIndex: u8, rest: []const u8, tokenType: TokenType) TokenType {
        const sourceSlice = self.start[startIndex .. startIndex + rest.len];

        return if (std.mem.eql(u8, sourceSlice, rest)) tokenType else .Identifier;
    }

    fn identifierType(self: *Scanner) TokenType {
        return switch (self.start[0]) {
            'a' => self.checkKeyword(1, "nd", .And),
            'c' => self.checkKeyword(1, "lass", .Class),
            'e' => self.checkKeyword(1, "lse", .Else),
            'f' => {
                if (self.start[0..self.current].len > 1) {
                    return switch (self.start[1]) {
                        'a' => self.checkKeyword(2, "lse", .False),
                        'o' => self.checkKeyword(2, "r", .For),
                        'n' => .Fn,
                        else => .Identifier,
                    };
                } else {
                    return .Identifier;
                }
            },
            'i' => self.checkKeyword(1, "f", .If),
            'n' => self.checkKeyword(1, "il", .Nil),
            'o' => self.checkKeyword(1, "r", .Or),
            'p' => self.checkKeyword(1, "rint", .Print),
            'r' => self.checkKeyword(1, "eturn", .Return),
            's' => self.checkKeyword(1, "uper", .Super),
            't' => {
                if (self.start[0..self.current].len > 1) {
                    return switch (self.start[1]) {
                        'h' => self.checkKeyword(2, "is", .This),
                        'r' => self.checkKeyword(2, "ue", .True),
                        else => .Identifier,
                    };
                } else {
                    return .Identifier;
                }
            },
            'v' => self.checkKeyword(1, "ar", .Var),
            'w' => self.checkKeyword(1, "hile", .While),
            else => .Identifier,
        };
    }

    fn skipWhitespace(self: *Scanner) void {
        while (true) {
            const char = self.peek();
            switch (char) {
                ' ', '\r', '\t' => self.advance(),
                '\n' => {
                    self.line += 1;
                    self.advance();
                },
                '/' => {
                    if (self.peekNext() == '/') {
                        // Comments go until the end of a line.
                        while (self.peek() != '\n' and !self.isAtEnd()) self.advance();
                    } else {
                        return;
                    }
                },
                else => return,
            }
        }
    }
};

fn isAlpha(char: u8) bool {
    return (char >= 'a' and char <= 'z') or
           (char >= 'A' and char <= 'Z') or
           char == '_';
}

fn isDigit(char: u8) bool {
    return char >= '0' and char <= '9';
}
